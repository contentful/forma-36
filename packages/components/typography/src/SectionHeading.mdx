---
title: 'SectionHeading'
type: 'component'
status: 'stable'
section: 'typographyComponents'
slug: /components/section-heading/
github: 'https://github.com/contentful/forma-36/tree/forma-v4/packages/components/typography/src'
typescript: ./SectionHeading.tsx
---

`SectionHeading` is intended to be used as a title for each section. SectionHeading are very important for accessibility, as they help to make copy more easily scannable and guides a user through your content. When using screen readers, it finds sections in your copy based on your heading hierarchy.

### Import

```js static=true
import { SectionHeading } from '@contentful/f36-components';
```

## Examples

### Basic

```jsx file=../examples/SectionHeadingExample.tsx

```

### Margins

By default `SectionHeading` component has a margin bottom of `spacingL` (`1.5rem`). You can override the default margin bottom by specifying it explicitly on the component level by `marginBottom="none"` or any other value from [our spacing options](/tokens/spacing)

```jsx file=../examples/SectionHeadingSpacingExample.tsx

```

### `as` property

By default `SectionHeading` has an `h2` tag. But you can change it be providing `as` prop with values from `h1` to `h6`.

```jsx file=../examples/SectionHeadingTagExample.tsx

```

### Truncated text

Sometimes you might have to truncate the text in the `SectionHeading` component, so we provide the prop especially for that. Under the hood, what it does is add `overflow: hidden` and set `text-overflow` to `ellipsis`. Have a look at the example below:

```jsx file=../examples/SectionHeadingTruncatedExample.tsx

```

### Composition

```jsx file=../examples/SectionHeadingCompositionExample.tsx

```

## Props (API reference)

<Props
  of="SectionHeading"
  storybookPath="/story/typography-section-heading--basic"
/>
