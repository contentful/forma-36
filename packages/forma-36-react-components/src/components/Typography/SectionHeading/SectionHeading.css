@import 'settings/colors';
@import 'settings/typography';
@import 'settings/transitions';

.SectionHeading {
  display: block;
  font-family: var(--font-stack-primary);
  font-weight: var(--font-weight-demi-bold);
  color: var(--color-text-dark);
  font-size: var(--font-size-s);
  line-height: var(--line-height-s);
  letter-spacing: var(--letter-spacing-wide);
  text-transform: uppercase;
  text-rendering: optimizeLegibility;
  margin: 0;
}
