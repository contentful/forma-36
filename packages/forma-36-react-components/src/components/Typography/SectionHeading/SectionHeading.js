import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import styles from './SectionHeading.css';

class SectionHeading extends React.Component {
  static propTypes = {
    extraClassNames: PropTypes.string,
    children: PropTypes.node.isRequired,
    testId: PropTypes.string,
    element: PropTypes.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'p']),
  };

  static defaultProps = {
    element: 'h3',
    extraClassNames: undefined,
    testId: 'cf-ui-section-heading',
  };

  render() {
    const {
      extraClassNames,
      children,
      testId,
      element,
      ...otherProps
    } = this.props;

    const classNames = cn(styles.SectionHeading, extraClassNames);

    const Element = element;

    return (
      <Element className={classNames} data-test-id={testId} {...otherProps}>
        {children}
      </Element>
    );
  }
}

export default SectionHeading;
