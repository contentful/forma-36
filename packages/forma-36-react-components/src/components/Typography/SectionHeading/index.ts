export * from './SectionHeading';
export { default } from './SectionHeading';
