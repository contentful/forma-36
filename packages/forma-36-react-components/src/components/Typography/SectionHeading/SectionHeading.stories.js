import React from 'react';
import { storiesOf } from '@storybook/react';
import { text, selectV2 } from '@storybook/addon-knobs/react';
import { host } from 'storybook-host';
import { withInfo } from '@storybook/addon-info';

import SectionHeading from './SectionHeading';

storiesOf('Components|Typography/SectionHeading', module)
  .addDecorator(
    host({
      align: 'center middle',
      cropMarks: false,
    }),
  )
  .add(
    'default',
    withInfo()(() => (
      <SectionHeading
        extraClassNames={text('Extra Class Names', '')}
        element={selectV2(
          'Element',
          ['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'p'],
          'h1',
        )}
      >
        {text('Content', 'Section heading')}
      </SectionHeading>
    )),
  );
