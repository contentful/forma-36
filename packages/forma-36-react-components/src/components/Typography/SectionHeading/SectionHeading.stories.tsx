import React from 'react';
import SectionHeading, { SectionHeadingProps } from './SectionHeading';

export default {
  title: 'Typography/SectionHeading',
  component: SectionHeading,
  parameters: {
    propTypes: [SectionHeading['__docgenInfo']],
  },
  argTypes: {
    className: { control: { disable: true } },
  },
};

interface Args extends SectionHeadingProps {
  sectionHeadingText: string;
}

export const Basic = ({ sectionHeadingText, ...args }: Args) => (
  <SectionHeading {...args}>{sectionHeadingText}</SectionHeading>
);

Basic.args = {
  sectionHeadingText: 'Section heading',
};
