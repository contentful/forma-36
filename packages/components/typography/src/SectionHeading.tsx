import React, { useContext } from 'react';
import tokens from '@contentful/f36-tokens';
import { css, cx } from 'emotion';
import {
  PolymorphicComponent,
  PolymorphicComponentWithRef,
} from '@contentful/f36-core';
import type { HeadingInternalProps, HeadingProps } from './Heading';
import { Heading } from './Heading';
import { TypographyContext } from './Typography';

const DEFAULT_TAG = 'h3';

export type SectionHeadingProps<E extends React.ElementType> = HeadingProps<E>;

const _SectionHeading: PolymorphicComponentWithRef<
  HeadingInternalProps,
  typeof DEFAULT_TAG
> = ({ children, className, ...otherProps }, ref) => {
  const configuration = useContext(TypographyContext);
  return (
    <Heading
      as={DEFAULT_TAG}
      testId="cf-ui-section-heading"
      marginBottom={configuration.sectionHeading}
      className={cx(
        css({
          fontSize: tokens.fontSizeS,
          lineHeight: tokens.lineHeightS,
          letterSpacing: tokens.letterSpacingWide,
          textTransform: 'uppercase',
        }),
        className,
      )}
      {...otherProps}
      ref={ref}
    >
      {children}
    </Heading>
  );
};

export const SectionHeading: PolymorphicComponent<
  HeadingInternalProps,
  typeof DEFAULT_TAG
> = React.forwardRef(_SectionHeading);
