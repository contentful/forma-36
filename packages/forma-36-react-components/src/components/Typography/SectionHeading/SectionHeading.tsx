import React from 'react';
import cn from 'classnames';
import styles from './SectionHeading.css';

import { TypographyContext } from '../Typography';

export interface SectionHeadingProps {
  element?: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'p';
  className?: string;
  children?: React.ReactNode | string;
  testId?: string;
  style?: React.CSSProperties;
}

export function SectionHeading({
  className,
  children,
  testId = 'cf-ui-section-heading',
  element = 'h3',
  ...otherProps
}: SectionHeadingProps): React.ReactElement {
  const classNames = cn(styles['SectionHeading'], className);

  const Element = element!; // eslint-disable-line @typescript-eslint/no-non-null-assertion

  return (
    <TypographyContext.Consumer>
      {(value) => {
        return (
          <Element
            className={cn(classNames, [
              value['sectionHeading'] &&
                `f36-margin-bottom--${value['sectionHeading']['spacing']}`,
            ])}
            data-test-id={testId}
            {...otherProps}
          >
            {children}
          </Element>
        );
      }}
    </TypographyContext.Consumer>
  );
}

export default SectionHeading;
