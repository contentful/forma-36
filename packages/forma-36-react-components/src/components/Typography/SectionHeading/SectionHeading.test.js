import React from 'react';
import { mount } from 'enzyme';
import { axe } from 'jest-axe';
import SectionHeading from './SectionHeading';

it('renders the component', () => {
  const output = mount(<SectionHeading>SectionHeading</SectionHeading>);

  expect(output).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const output = mount(
    <SectionHeading className="my-extra-class">SectionHeading</SectionHeading>,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component h3', () => {
  const output = mount(
    <SectionHeading element="h3">SectionHeading</SectionHeading>,
  );

  expect(output).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const output = mount(<SectionHeading>SectionHeading</SectionHeading>).html();
  const results = await axe(output);

  expect(results).toHaveNoViolations();
});
