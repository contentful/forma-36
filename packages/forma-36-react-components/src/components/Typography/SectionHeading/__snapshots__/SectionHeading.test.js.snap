// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<h3
  className="SectionHeading"
  data-test-id="cf-ui-section-heading"
>
  SectionHeading
</h3>
`;

exports[`renders the component h3 1`] = `
<h3
  className="SectionHeading"
  data-test-id="cf-ui-section-heading"
>
  SectionHeading
</h3>
`;

exports[`renders the component with an additional class name 1`] = `
<h3
  className="SectionHeading my-extra-class"
  data-test-id="cf-ui-section-heading"
>
  SectionHeading
</h3>
`;
